src/paths/floyd_warshall.c

```diff
@@ -140,3 +142,132 @@ igraph_error_t igraph_distances_floyd_warshall(
+
+    for (igraph_integer_t v=0; v < no_of_nodes; v++) {
+        MATRIX(*res, v, v) = 0;
+    }
+
+    for (igraph_integer_t e=0; e < no_of_edges; e++) {
+        igraph_integer_t from = IGRAPH_FROM(graph, e);
+        igraph_integer_t to = IGRAPH_TO(graph, e);
+        igraph_real_t w = weights ? VECTOR(*weights)[e] : 1;
+
+        if (w < 0) {
+            if (mode == IGRAPH_ALL) {
+                IGRAPH_ERRORF("Negative edge weight (%g) found in undirected graph "
+                              "while calculating distances with Floyd-Warshall.",
+                              IGRAPH_ENEGLOOP, w);
+            } else if (to == from) {
+                IGRAPH_ERRORF("Self-loop with negative weight (%g) found "
+                              "while calculating distances with Floyd-Warshall.",
+                              IGRAPH_ENEGLOOP, w);
+            }
+        }
+
+        if (out && MATRIX(*res, from, to) > w) MATRIX(*res, from, to) = w;
+        if (in  && MATRIX(*res, to, from) > w) MATRIX(*res, to, from) = w;
+    }
+
+    igraph_matrix_int_t predecessors;
+    igraph_stack_int_t stack;
+
+    IGRAPH_STACK_INT_INIT_FINALLY(&stack, no_of_nodes);
+
+    IGRAPH_CHECK(igraph_matrix_int_init(&predecessors, no_of_nodes, no_of_nodes));
+    IGRAPH_FINALLY(igraph_matrix_int_destroy, &predecessors);
+
+    for (igraph_integer_t v=0; v < no_of_nodes; v++) {
+        for (igraph_integer_t u=0; u < no_of_nodes; u++) {
+            /* the penultimate vertex on the shortest path from u to v */
+            MATRIX(predecessors, u, v) = u;
+        }
+    }
+
+    igraph_t out_k;
+    IGRAPH_CHECK(igraph_empty(&out_k, 0, IGRAPH_DIRECTED));
+    IGRAPH_FINALLY(igraph_destroy, &out_k);
+    igraph_vector_int_t children;
+    IGRAPH_VECTOR_INT_INIT_FINALLY(&children, 0);
+    igraph_vector_int_t parents;
+    IGRAPH_VECTOR_INT_INIT_FINALLY(&parents, no_of_nodes);
+
+    for (igraph_integer_t k=0; k < no_of_nodes; k++) {
+        for (igraph_integer_t v=0; v < no_of_nodes; v++) {
+            VECTOR(parents)[v] = v == k ? -1 : MATRIX(predecessors, k, v);
+        }
+        IGRAPH_CHECK(igraph_tree_from_parent_vector(&out_k, &parents, IGRAPH_TREE_OUT));
+         /* Iteration order matters for performance!
+          * First j, then i, because matrices are stored as column-major. */
+        for (igraph_integer_t j=0; j < no_of_nodes; j++) {
+            igraph_real_t dkj = MATRIX(*res, k, j);
+            if (dkj == IGRAPH_INFINITY) continue;
+            IGRAPH_CHECK(igraph_stack_int_push(&stack, k));
+            while (!igraph_stack_int_empty(&stack)) {
+                igraph_integer_t v = igraph_stack_int_pop(&stack);
+                igraph_neighbors(&out_k, &children, v, IGRAPH_OUT);
+                igraph_integer_t no_of_children = igraph_vector_int_size(&children);
+                for (igraph_integer_t l=0; l < no_of_children; l++) {
+                    igraph_integer_t i = VECTOR(children)[l];
```

Comment:
There was a mistake in my previous version.  In fact, we need the `children`, not only their count `no_of_children `. 
It is strange that in spite of that mistake all the tests were passing. 
Also looking into this closer, the algorithm shouldn't work for undirected graphs at this point.